model/perf.go

```diff
@@ -637,3 +655,174 @@ func (r *PerformanceResults) FindOutdatedRollups(ctx context.Context, name strin
+		{
+			"$match": bson.M{
+				"info.order": bson.M{
+					"$exists": true,
+				},
+				"info.mainline":  true,
+				"info.project":   bson.M{"$exists": true},
+				"info.variant":   bson.M{"$exists": true},
+				"info.task_name": bson.M{"$exists": true},
+				"info.test_name": bson.M{"$exists": true},
+			},
+		},
+		{
+			"$unwind": "$rollups.stats",
+		},
+		{
+			"$group": bson.M{
+				"_id": bson.M{
+					"task":        "$info.task_name",
+					"variant":     "$info.variant",
+					"project":     "$info.project",
+					"test":        "$info.test_name",
+					"measurement": "$rollups.stats.name",
+				},
+			},
+		},
+		{
+			"$replaceRoot": bson.M{
+				"newRoot": "$_id",
+			},
+		},
+	}
+	cur, err := env.GetDB().Collection(perfResultCollection).Aggregate(ctx, pipeline)
+	if err != nil {
+		return nil, errors.Wrap(err, "Cannot aggregate time series ids")
+	}
+	defer cur.Close(ctx)
+	var res []TimeSeriesId
+	err = cur.All(ctx, &res)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not decode time series ids")
+	}
+	return res, nil
+}
+
+type TimeSeriesId struct {
+	Project     string `bson:"project"`
+	Variant     string `bson:"variant"`
+	Task        string `bson:"task"`
+	Test        string `bson:"test"`
+	Measurement string `bson:"measurement"`
+}
+
+type TimeSeriesEntry struct {
+	PerfResultID string  `bson:"_id"`
+	Value        float64 `bson:"value"`
+	Order        int     `bson:"order"`
+}
+
+type TimeSeries struct {
+	TimeSeriesId TimeSeriesId      `bson:"_id"`
+	Data         []TimeSeriesEntry `bson:"times_series"`
+}
+
+func makeTimeSeriesPipeline(timeSeriesId TimeSeriesId) []bson.M {
+	return []bson.M{
+		{
+			"$match": bson.M{
+				"info.project":   timeSeriesId.Project,
+				"info.variant":   timeSeriesId.Variant,
+				"info.task_name": timeSeriesId.Task,
+				"info.test_name": timeSeriesId.Test,
+			},
+		},
+		{
+			"$match": bson.M{
+				"info.order": bson.M{
+					"$exists": true,
+				},
+				"info.mainline": true,
+			},
+		},
+		{
+			"$unwind": "$rollups.stats",
+		},
+		{
+			"$match": bson.M{
+				"rollups.stats.name": timeSeriesId.Measurement,
+			},
+		},
+		{
+			"$project": bson.M{
+				"value": "$rollups.stats.val",
+				"order": "$info.order",
+			},
+		},
+	}
+}
+
+func GetTimeSeries(ctx context.Context, env cedar.Environment, timeSeriesId TimeSeriesId) (*TimeSeries, error) {
+	timeSeriesAggregator := makeTimeSeriesPipeline(timeSeriesId)
+	cur, err := env.GetDB().Collection(perfResultCollection).Aggregate(ctx, timeSeriesAggregator)
+	if err != nil {
+		return nil, errors.Wrap(err, "Cannot aggregate time series")
+	}
+	defer cur.Close(ctx)
+	var res []TimeSeriesEntry
+	err = cur.All(ctx, &res)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not decode time series")
+	}
+	return &TimeSeries{
+		TimeSeriesId: timeSeriesId,
+		Data:         res,
+	}, nil
+}
+
+func ReplaceChangePoints(ctx context.Context, env cedar.Environment, timeSeries *TimeSeries, changePoints []ChangePoint) []error {
```

Comment:
please don't return a slice of errors, use `grip.NewBasicCatcher` and then the methods on that interface to collect errors.

Reply:
Ok